Add "dataTableHeadersMissing" audit rule.

atomanyih opened the pull request:
This rule tests for compliance with [Section 508 #1194.22g](http://www.section508.gov/content/learn/standards/quick-reference-guide#1194.22g)

We also added [an entry to the wiki](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_table_01)

@matt-royal

matt-royal commented:
We're waiting to hear back from our legal team about the CLA. We should be able to sign it early next week.

atomanyih commented:
I signed it!

atomanyih commented:
Any word on this pull request?

RickSBrown commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,67 @@
+goog.require('axs.constants.Severity');
+
+(function() {
+  function tableDoesNotHaveHeaderRow(rows) {
```
Style nit: tab indent is 4 spaces not two

RickSBrown commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,67 @@
+goog.require('axs.AuditRules');
```
Missing licence comment at start (copy it from another audit and update the date to 2015)

RickSBrown commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,67 @@
+    test: function(element) {
+
+      if (element.getAttribute('role') == 'presentation') {
+        return element.querySelectorAll('th').length != 0
```
nit: semicolon

RickSBrown commented:
Sorry nobody is full-time on this.
I've done a quick first pass looking for obvious nits which will speed things up when Alice does get to this.
Also please add a changelog entry - this would obviously go under `Enhancements`

RickSBrown commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+  var table = document.createElement('table');
+
+  rows.forEach(function(row) {
+    table.appendChild(buildRow(row))
```
semicolon

RickSBrown commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+    row.appendChild(document.createElement(cell));
+  });
+
+  return row
```
semicolon

RickSBrown commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+  var tbody = document.createElement('tbody');
+
+  bodyRows.forEach(function(row) {
+    tbody.appendChild(buildRow(row))
```
semicolon

atomanyih commented:
@RickSBrown Thanks for the feedback! We'll go over it and make those changes

atomanyih commented:
@RickSBrown Addressed the comments and added an entry to the changelog. I put it under 'New rules' instead of enhancements because that seemed like a better fit.

RickSBrown commented:
Oh yeah, 'New rules', good call.

alice commented:
Apologies for my extreme slowness on this! Looking now.

alice commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,81 @@
+goog.require('axs.constants.Severity');
+
+(function () {
+    function tableDoesNotHaveHeaderRow(rows) {
```
Would you be able to move these three methods into AccessibilityUtils as something like 
```
/**
 * @param {HTMLTableElement} table
 * @return {Array<HTMLTableRowElement>}
 */
axs.utils.headerRowsForTable = function(table) {
    // ...
};

/**
 * @param {HTMLTableElement} table
 * @return {Array<HTMLTableColumnElement>}
 */
axs.utils.headerColumnsForTable = function(table) {
    // ...
};

/**
 * @param {HTMLTableElement} table
 * @return {boolean}
 */
axs.utils.isDataTable = function(table) {
    //...
};

atomanyih replied:
The only reason we didn't move it out is because we're the only ones using it and we saw other tests with private helpers. If you think these methods might have more wide use, we can totes move them out

alice replied:
Yeah, my preference is for things to go in axs.utils unless it really can only apply to the current test, since things might be useful for the debugging UI in the extension. (In fact, https://github.com/GoogleChrome/accessibility-developer-tools/blob/master/src/audits/RequiredOwnedAriaRoleMissing.js#L75 should probably move into axs.utils as well at some point.)

That said, perhaps these methods are single purpose enough that they can stay here for the time being - could you just JSDoc them instead?

alice commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,81 @@
+        return false;
+    }
+
+    function tableDoesNotHaveGridLayout(rows) {
```
Does this do anything other than the union of the previous two methods?

atomanyih replied:
Grid layout is a little different because you don't have to have a header in the top left cell

alice commented on src/audits/DataTableHeadersMissing.js:
```diff
@@ -0,0 +1,81 @@
+        url: 'https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_table_01',
+        severity: axs.constants.Severity.SEVERE,
+        relevantElementMatcher: function (element) {
+            return axs.browserUtils.matchSelector(element, 'table');
```
Either this should select for only data tables, or else the heading of this rule is confusing. I see you check for `<th>` on non-data tables below; perhaps this rule should be non-opinionated about the non-data table case, and exclude them here?

atomanyih replied:
That's a good point, we'll change that

ctaymor replied:
Upon revisiting, remembered that this rule was based on the [WebAim interpretation of 508g](http://webaim.org/standards/508/checklist) which says that presentation tables should not have any headers at all. We updated the wiki explanation and the audit name to clarify this.

alice commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+module('DataTableHeadersMissing');
+
+function buildRow(cells) {
```
This is slick, but could you:
1) Scope using an IIFE
1) JSDoc these methods (like https://github.com/GoogleChrome/accessibility-developer-tools/blob/master/test/audits/multiple-aria-owners-test.js#L18)

alice commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+    deepEqual(actual.elements, [table]);
+});
+
+test('Table uses a grid layout', function () {
```
Ahhh now I get what that grid layout method does :)

atomanyih replied:
Oh hi

alice commented on test/audits/data-table-headers-missing-test.js:
```diff
@@ -0,0 +1,285 @@
+    return table;
+}
+
+test('Table with a header row', function () {
```
Thanks for all these awesome tests!

alice commented:
Added a bunch of comments as you saw - apologies again for being so slow, but looking forward to hopefully getting this merged!

ctaymor commented:
@alice We addressed your comments. Let us know what you think about the rename. We were wavering on pulling the presentational table check out as a separate test, but since there weren't test utils, we decided to just rename.


@ctaymor @atomanyih & The Pivotal UI Team
@d-reinhold @kennyw12 @matt-royal @nicw @stubbornella

alice commented on Changelog.md:
```diff
@@ -6,6 +6,7 @@
 ### New rules
 
 * A label element may not have labelable descendants other than its labeled control (`src/audits/MultipleLabelableElementsPerLabel.js`)
+* A data table must identify row and column headers (`src/audits/TableHasAppropriateHeaders.js`)
```
Sorry, because I took a while to get back to this, this Changelog line now needs to be moved above the last Release heading.

alice commented:
:+1: once Changelog updated.

kennyw12 commented:
Sorry I tried rebasing and made a mess.

kennyw12 commented:
It's all good to go.

jdan commented on Changelog.md:
```diff
@@ -1,5 +1,7 @@
 ### New rules
 * A tabpanel should be related to a tab via aria-controls or aria-labelledby (`src/audits/UncontrolledTabpanel.js)
+* A data table must identify row and column headers (`src/audits/TableHasAppropriateHeaders.js`)
+* A tabpanel should be related to a tab via aria-controls or aria-labelledby (`src/audits/UncontrolledTabpanel.js)
```
Unmatched backtick here :)

ctaymor replied:
The backtick is missing, but also looks like some new rules got duplicated here in the changelog merge. I've fixed both. :)

atomanyih commented:
@alice Anything keeping this from being merged?

alice commented:
My apologies for my slowness; looking now, will ping with any issues or else merge it :)

alice commented:
Merging, thanks for your patience!

atomanyih commented:
@alice No worries, thanks for merging it in! 

Any chance you could take a look at https://github.com/GoogleChrome/accessibility-developer-tools/pull/233 as well?
